From: Vinayak Menon <vinmenon@codeaurora.org>
Subject: [PATCH 2/2] mm: vmpressure: fix sending wrong events on underflow
Date: Fri, 27 Jan 2017 13:43:37 +0530
Cc: linux-mm@kvack.org,
	linux-kernel@vger.kernel.org,
	Vinayak Menon <vinmenon@codeaurora.org>
To: akpm@linux-foundation.org,
	hannes@cmpxchg.org,
	mgorman@techsingularity.net,
	vbabka@suse.cz,
	mhocko@suse.com,
	riel@redhat.com,
	vdavydov.dev@gmail.com,
	anton.vorontsov@linaro.org,
	minchan@kernel.org,
	shashim@codeaurora.org

At the end of a window period, if the reclaimed pages
is greater than scanned, an unsigned underflow can
result in a huge pressure value and thus a critical event.
Reclaimed pages is found to go higher than scanned because
of the addition of reclaimed slab pages to reclaimed in
shrink_node without a corresponding increment to scanned
pages. Minchan Kim mentioned that this can also happen in
the case of a THP page where the scanned is 1 and reclaimed
could be 512.

Signed-off-by: Vinayak Menon <vinmenon@codeaurora.org>
---
 mm/vmpressure.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/mm/vmpressure.c b/mm/vmpressure.c
index 149fdf6..3281b34 100644
--- a/mm/vmpressure.c
+++ b/mm/vmpressure.c
@@ -112,8 +112,10 @@ static enum vmpressure_levels vmpressure_calc_level(unsigned long scanned,
 						    unsigned long reclaimed)
 {
 	unsigned long scale = scanned + reclaimed;
-	unsigned long pressure;
+	unsigned long pressure = 0;
 
+	if (reclaimed >= scanned)
+		goto out;
 	/*
 	 * We calculate the ratio (in percents) of how many pages were
 	 * scanned vs. reclaimed in a given time frame (window). Note that
@@ -124,6 +126,7 @@ static enum vmpressure_levels vmpressure_calc_level(unsigned long scanned,
 	pressure = scale - (reclaimed * scale / scanned);
 	pressure = pressure * 100 / scale;
 
+out:
 	pr_debug("%s: %3lu  (s: %lu  r: %lu)\n", __func__, pressure,
 		 scanned, reclaimed);
 
